File: Path Sum III/Mine.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
struct pair_hash
{
    template <class T1, class T2>
    size_t operator() (const pair<T1, T2> &pair) const
    {
        return hash<T1>()(pair.first) ^ hash<T2>()(pair.second);
    }
};
class Solution {
private:
    int result = 0;
    int sum1;
public:
    void calculate_sum (TreeNode* root, int current_sum){
        current_sum = current_sum + root -> val;
        if (current_sum == sum1)
            result++;
        if (root->left)
            calculate_sum (root->left, current_sum);
        if (root->right)
            calculate_sum (root->right, current_sum);
    }
    void traversal (TreeNode* root){
        if (root)
            calculate_sum (root, 0);
        if (root->left)
            traversal(root->left);
        if (root->right)
            traversal(root->right);
            
    }
    int pathSum(TreeNode* root, int sum) {
        sum1 = sum;
        if (!root)
            return 0;
        traversal (root);
        return result;
    }
};
// Pure dfs, find all situations without optimization. AC

File: Problems/236. Lowest Common Ancestor of a Binary Tree/mine.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


# 最坏的情况 我们需要遍历所有的节点 所以O(N)
# DFS 会占用空间 所有空间也是 O(N)
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        lowest_common_ancestor = self.findlowestCommonAncestor(root, p, q)
        
        return lowest_common_ancestor
    
    def findlowestCommonAncestor(self, current, p, q):
        if not current:
            return None
        
        if current == p or current == q:
            return current
        
        left = self.findlowestCommonAncestor(current.left, p, q)
        right = self.findlowestCommonAncestor(current.right, p, q)
        
        if left and right:
            return current
        
        if left:
            return left
        
        if right:
            return right
        
        return None
